workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourManagerInitializationSystem.cs

```diff
@@ -0,0 +1,89 @@
+using System;
+using Unity.Collections;
+using Unity.Entities;
+using UnityEngine;
+
+namespace Improbable.Gdk.Core
+{
+    /// <summary>
+    ///     Creates and removes SpatialOSBehaviourManager object for EntityGameObjects.
+    /// </summary>
+    [UpdateInGroup(typeof(SpatialOSReceiveGroup.GameObjectInitializationGroup))]
+    public class SpatialOSBehaviourManagerInitializationSystem : ComponentSystem
+    {
+        public struct AddedEntitiesData
+        {
+            public readonly int Length;
+            public EntityArray Entities;
+            public ComponentArray<GameObjectReference> GameObjectReferences;
+            [ReadOnly] public ComponentDataArray<GameObjectReferenceHandle> GameObjectReferenceHandles;
+            [ReadOnly] public ComponentDataArray<RequiresSpatialOSBehaviourManager> RequiresSpatialOSBehaviourManagerTags;
+        }
+
+        public struct RemovedEntitiesData
+        {
+            public readonly int Length;
+            public EntityArray Entities;
+            [ReadOnly] public ComponentDataArray<GameObjectReferenceHandle> GameObjectReferenceHandles;
+            public SubtractiveComponent<GameObjectReference> NoGameObjectReference;
+        }
+
+        [Inject] private AddedEntitiesData addedEntitiesData;
+        [Inject] private RemovedEntitiesData removedEntitiesData;
+
+        private GameObjectDispatcherSystem gameObjectDispatcherSystem;
+        private ILogDispatcher logDispatcher;
+
+        protected override void OnCreateManager(int capacity)
+        {
+            base.OnCreateManager(capacity);
+
+            gameObjectDispatcherSystem = World.GetExistingManager<GameObjectDispatcherSystem>();
+            if (gameObjectDispatcherSystem == null)
+            {
+                throw new GameObjectDispatcherSystemNotFoundException("GameObjectDispatcherSystem not found.");
+            }
+
+            logDispatcher = WorkerRegistry.GetWorkerForWorld(World).View.LogDispatcher;
+        }
+
+        protected override void OnUpdate()
+        {
+            for (var i = 0; i < addedEntitiesData.Length; i++)
+            {
+                var entityIndex = addedEntitiesData.Entities[i].Index;
+                if (gameObjectDispatcherSystem.EntityIndexToSpatialOSBehaviourManager.ContainsKey(entityIndex))
+                {
+                    logDispatcher.HandleLog(LogType.Error, new LogEvent(
+                            "Entity already has a SpatialOSBehaviourManager.")
+                        .WithField("EntityIndex", entityIndex));
+                    continue;
+                }
+
+                var gameObject = addedEntitiesData.GameObjectReferences[i].GameObject;
+                gameObjectDispatcherSystem.EntityIndexToSpatialOSBehaviourManager[entityIndex] = new SpatialOSBehaviourManager(gameObject);
```

Comment:
Can combine the two lines here

Reply:
Isn't it more readable as it is though? And I'd expect a decent compiler to avoid performance overhead here as well.

Reply:
Done!